### DISCUSSION TITLE
Fragment Variables with `client-preset`

### DISCUSSION THREAD

**Discussion by nandorojo** [dotansimha/graphql-code-generator#8603] [Q&A] [answered]
Hey there,

I'm a fan of the new `client-preset` + Relay-style querying. I'm trying to adopt the recommended approach when building screens. However, I don't see a way to use variables with fragments.

Take the example as shown in the [blog post](https://www.the-guild.dev/blog/unleash-the-power-of-fragments-with-graphql-codegen): 

![image](https://user-images.githubusercontent.com/13172299/201416541-c8401759-bb20-4288-bcb6-855cf21e0a98.png)

In the third example (`<FriendsList />`), the query has variables of `first: 5`. But more commonly, root queries have variables defined outside. Is there a way to implement variables with fragments in the client preset?

**Comment by saihaj** [comment, marked as answer]
Currently fragment arguments are not possible as per spec see RFC https://github.com/graphql/graphql-spec/pull/865
Relay has [custom directives](https://relay.dev/docs/api-reference/graphql-and-directives/#arguments) which get converted by compiler  we could look into doing a similar thing 🤔 cc @n1ru4l

**Reply by n1ru4l** [reply]
I think we could make fragment arguments possible with a custom parser and transforms.

**Reply by nandorojo** [reply]
That would be awesome.

**Reply by nandorojo** [reply]
Out of curiosity, how likely is it that the custom directives could get added? I know that's a tough question with open source. I wish I could help with a PR, but this isn't my area of expertise. 

I think it would help a lot for lists using client preset. Thanks again guys!

**Reply by nandorojo** [reply]
I see that `graphql-js` supports fragment variables AST behind an experimental flag: https://github.com/graphql/graphql-spec/pull/865/files#diff-d8f3e7a05551a43d75f21b0cfcd2a64b8910bc37be8b221dc069b2986bda575fR91

**Reply by nandorojo** [reply]
I see that GraphQL codegen's TypeScript config allows this: https://the-guild.dev/graphql/codegen/plugins/typescript/typescript-operations#experimentalfragmentvariables

Is it possible to add a similar flag to client preset?

**Reply by nandorojo** [reply]
Linking to this for anyone wondering about fragment variable support: https://github.com/dotansimha/graphql-yoga/issues/3209

**Reply by osdiab** [reply]
Linking to this for anyone wondering [what the fragment variables syntax looks like](https://github.com/dotansimha/graphql-code-generator/blob/ac778ce6533ddeb9a40ecb4f9764a50a9a9bad0e/packages/plugins/typescript/operations/tests/ts-documents.spec.ts#L329) as `graphql-codegen` docs don't seem to describe it anywhere

**Reply by osdiab** [reply]
Also, this is the current state of the fragment variables feature in the spec - it looks like there's active development happening now so hopefully it goes somewhere! https://github.com/graphql/graphql-spec/pull/1081
